### DISCUSSION TITLE
Windows loses connection to IKEv2 server after some time (despite reporting it is connected in system tray)

### DISCUSSION THREAD

**Discussion by acrolink** [strongswan/strongswan#639] [Q&A]
I have a strange error in one Windows 10 machine (latest build). It connects to the IKEv2 server using client certificate. Connects successfully but after some time (15 minutes or so) the connection is lost. In the meantime, Windows reports that it is connected in the status bar. Solution is to disconnect and connect again.

My question, what might cause this situation? I am sure it is not server side related since other clients keep connected.

Can it be an issue with `Cipher Suites` used? Which ones are more reliable to keep the connection online? 

What to check for in that system?

Can it be caused by the Internet Service Provider? or by the router installed there?

Here is a sample log entry:

```
Sep 29 06:27:24 localhost charon-systemd[4993]: initiating IKE_SA win7-client-a[71] to x.x.x.x
Sep 29 06:27:24 localhost charon-systemd[4993]: generating CREATE_CHILD_SA request 2 [ SA No KE ]
Sep 29 06:27:24 localhost charon-systemd[4993]: sending packet: from my.server.ip.addr[4500] to x.x.x.x[1141] (572 bytes)
Sep 29 06:27:24 localhost charon-systemd[4993]: received packet: from x.x.x.x[1141] to my.server.ip.addr[4500] (444 bytes)
Sep 29 06:27:24 localhost charon-systemd[4993]: parsed CREATE_CHILD_SA response 2 [ SA KE No ]
Sep 29 06:27:24 localhost charon-systemd[4993]: selected proposal: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
Sep 29 06:27:24 localhost charon-systemd[4993]: scheduling rekeying in 14385s
Sep 29 06:27:24 localhost charon-systemd[4993]: maximum IKE_SA lifetime 15825s
Sep 29 06:27:24 localhost charon-systemd[4993]: IKE_SA win7-client-a[71] rekeyed between my.server.ip.addr[my.ikev2.server]...x.x.x.x[C=XX, O=XX, CN=client.a@my.ikev2.server]
Sep 29 06:27:24 localhost charon-systemd[4993]: deleting IKE_SA win7-client-a[64] between my.server.ip.addr[my.ikev2.server]...x.x.x.x[C=XX, O=XX, CN=client.a@my.ikev2.server]
Sep 29 06:27:24 localhost charon-systemd[4993]: sending DELETE for IKE_SA win7-client-a[64]
Sep 29 06:27:24 localhost charon-systemd[4993]: generating INFORMATIONAL request 3 [ D ]
Sep 29 06:27:24 localhost charon-systemd[4993]: sending packet: from my.server.ip.addr[4500] to x.x.x.x[1141] (76 bytes)
Sep 29 06:27:24 localhost charon-systemd[4993]: received packet: from x.x.x.x[1141] to my.server.ip.addr[4500] (76 bytes)
Sep 29 06:27:24 localhost charon-systemd[4993]: parsed INFORMATIONAL response 3 [ ]
Sep 29 06:27:24 localhost charon-systemd[4993]: IKE_SA deleted
```

**Comment by tobiasbrunner** [comment]
The log shows an IKE_SA rekeying initiated by the server. It seems to complete successfully. Is this actually when the connection fails? Because this shouldn't affect the CHILD_SAs that transport the traffic. A complete log might help (from the initial connection until it actually fails, see #196 for log settings).

Some Windows clients definitely have/had issues with rekeying (see e.g. [this older ticket](https://wiki.strongswan.org/issues/3400), although the workaround there was to let the server initiate the rekeying).

**Comment by acrolink** [comment]
@tobiasbrunner Thanks for the valuable information.

I read the following from the old ticket you referred to:

_The bug seems to happen because Windows proposes different algorithms during IKE rekeying than it did during the initial connection; it then doesn't successfully complete IKE rekeying if the integrity algorithm changes as a result (even though the server thinks it rekeyed successfully)._

I have just observed the same and aligned the algorithms accordingly. I will surely report back here the result.

**Comment by acrolink** [comment]
Here is what I have observed:

Windows upon initializing the connection proposed:

`ESP:AES_CBC_128/HMAC_SHA1_96/NO_EXT_SEQ`

On rekeying it requested in addition `MODP_768` like this:

`ESP:AES_CBC_128/HMAC_SHA1_96/MODP_768/NO_EXT_SEQ`

So what I have done is to configure the `swanctl.conf` child like this:

`esp_proposals = aes128-sha1-modp768,aes192gcm16-aes128gcm16-prfsha256-ecp256-modp3072,aes192-sha256-ecp256-modp3072,default`

Worth mentioning that I have also configured the allowed ciphers on Windows side like this:

`Set-VpnConnectionIPsecConfiguration -ConnectionName "My VPN" -AuthenticationTransformConstants SHA196 -CipherTransformConstants AES128 -DHGroup Group14 -EncryptionMethod AES128 -IntegrityCheckMethod SHA256 -PFSgroup PFS1 -Force` (powershell).

@tobiasbrunner , Do you think `ESP:AES_CBC_128/HMAC_SHA1_96/MODP_768/NO_EXT_SEQ` is problematic from a security perspective?

The used `IKE` ciphers are `AES_CBC-256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048`.

**Reply by tobiasbrunner** [reply]
> Do you think `ESP:AES_CBC_128/HMAC_SHA1_96/MODP_768/NO_EXT_SEQ` is problematic from a security perspective?

Yes, definitely. The reason why this DH group is selected for the CHILD_SA rekeying, is this: `-PFSgroup PFS1` (DH group 1 is modp786). What you probably want there is `-PFSgroup PFS2048` (or `PFSMM`, which should default to the DH group used for the IKE_SA).

**Comment by acrolink** [comment]
Indeed, seems the issue was related to Windows asking for other cipher suites on rekeying than the ones used on connection initialization. Also, not all Windows 10 systems show this problem. Providing those cipher suites by the server or configuring Windows cipher suites to match those of the server solves the issue.

**Comment by acrolink** [comment]
@tobiasbrunner More information for future debugging and possible documentation updates:

I have noticed that `ESP: AES_CBC-128/HMAC_SHA2_256_128` for Windows 10 clients causes connection loss after couple of minutes, the solution is to use `ESP: AES_CBC-128/HMAC_SHA1_96`. I don't know what is the underlying cause.

For your consideration and thought. Thank again for the great work.

**Reply by tobiasbrunner** [reply]
Isn't that basically what Malcolm reported in the old ticket? (The switch of the integrity algorithm to SHA-1 during rekeying.)

**Reply by acrolink** [reply]
> Isn't that basically what Malcolm reported in the old ticket? (The switch of the integrity algorithm to SHA-1 during rekeying.)

Yes, indeed. Exactly as it is described there: `Windows apparently by default supports SHA256 for IKE, but only SHA1 for ESP`.

**Comment by acrolink** [comment]
This is also stable (with no connection drops):

`proposals = aes256-sha384-prfsha384-ecp384`
`esp_proposals = aes256gcm16-ecp384`

**IMPORTANT**: `NegotiateDH2048_AES256` must not be set in registry or at least, possibly, adjusted to allow these cipher suites.

**Reply by tobiasbrunner** [reply]
Good to know. AES-GCM is a good choice for ESP anyway, as it's more efficient than the classic combination of an encryption and integrity algorithm.

**Comment by hostalp** [comment]
When dealing with the same kind of issue as described here and especially at https://wiki.strongswan.org/issues/3400 - e.g.:

1. Windows 11 client on initial connection:
`received proposals: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048, IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048, IKE:AES_CBC_256/HMAC_SHA2_384_192/PRF_HMAC_SHA2_384/MODP_2048`
and selecting `selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048`
2 Windows 11 client on strongSwan-initiated IKE SA rekey:
`received proposals: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048`
and selecting that one

With strongSwan configured with:
`configured proposals: IKE:AES_GCM_16_128/AES_CCM_16_128/PRF_HMAC_SHA2_256/ECP_384/ECP_256/MODP_4096/MODP_2048, IKE:AES_GCM_16_256/AES_CCM_16_256/PRF_HMAC_SHA2_256/ECP_384/ECP_256/MODP_4096/MODP_2048, IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384/ECP_256/MODP_4096/MODP_2048, IKE:AES_GCM_16_128/AES_CCM_16_128/PRF_AES128_XCBC/ECP_384/ECP_256/MODP_4096/MODP_2048, IKE:AES_CBC_256/AES_XCBC_96/PRF_AES128_XCBC/ECP_384/ECP_256/MODP_4096/MODP_2048, IKE:AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048`

Which could be worked around by removing (at the strongSwan side) the last entry that allowed `AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048`, which isn't optimal because it was there for compatibility with other clients...

I was playing with an idea to configure the strongSwan to simply propose (on rekey) only the cipher suite that's been selected on initial connection and therefore already used so far. But there doesn't appear to be such option. Would that make sense to implement?
